Eslint - enforces many more rules.

ColinEberhardt opened the pull request:
Based on the review comments I saw for #138 I thought the linting rules could do with being 'tightened up'.

I checked the `.eslintrc` and found that the only thing being checked was the quotes style, and that was only being applied to `Layout.js`!

I've auto-configured ESLint based on the existing codebase, then added a few rules that I think are most important. Let me know what you think.

vjeux commented on src/Layout.js:
```diff
@@ -136,11 +136,11 @@ var computeLayout = (function() {
     }
 
-    if (value != null) {
+    if (value !== null && typeof value !== 'undefined') {
```
You should do

value !== undefined

The typeof check is when the variable may or may not exist. This is useful for checking globals. But in this case we know it exists so we can just check against the undefined value.

vjeux commented on src/Layout.js:
```diff
@@ -186,11 +186,11 @@ var computeLayout = (function() {
     }
 
-    if (value != null && value >= 0) {
+    if (value !== null && value >= 0) {
```
Probably want undefined here as well and not null

devongovett replied:
`!=` null covers both cases (`undefined == null`, but `undefined !== null`).

ColinEberhardt commented:
@vjeux @devongovett - I've relaxed the eqeqeq rule to permit eqeq checks against null, as @devongovett rightly points out this allows you to check for both null and undefined with a single test.

ColinEberhardt commented:
As an side, it does feel like the `getLeadingMargin` function, and its neighbours, share a lot of common code. I'll have a look at refactoring as a separate PR.

spicyj commented:
See how React and Relay have eslint set up – they use a common config from the fbjs-scripts package.

ColinEberhardt commented:
@spicyj thanks for the heads up, didn't realise there was a centralised lint configuration. I've updated the PR to use the `fbjs-scripts` eslint configuration and fixed all errors / warnings.

vjeux commented on src/Layout.js:
```diff
@@ -410,7 +410,7 @@ var computeLayout = (function() {
   function setDimensionFromStyle(node, axis) {
     // The parent already computed us a width or height. We just skip it
-    if (node.layout[dim[axis]] != null) {
+    if (typeof node.layout[dim[axis]] !== 'undefined') {
```
please use `x === undefined` instead of `typeof x === 'undefined'`

vjeux commented on src/Layout.js:
```diff
@@ -391,10 +391,10 @@ var computeLayout = (function() {
 
     var boundValue = value;
-    if (max != null && max >= 0 && boundValue > max) {
+    if (max !== null && max >= 0 && boundValue > max) {
```
should be `!== undefined`

vjeux commented:
Sweet! Outside of the typeof undefined and null, looks very good! Thanks for doing that

ColinEberhardt commented:
> please use x === undefined instead of typeof x === 'undefined'

Fair point, guarding against redefinition of `undefined` is a little unnecessary, and does make the code less readable. I've made these updates.

> Thanks for doing that

No problems :-)

vjeux commented:
Feel free to land if travis passes :)

ColinEberhardt commented:
Travis is happy - all done :+1:
